### ISSUE TITLE
Provide workaround for IAccessible implementations with broken next/previous object retrieval

### ISSUE THREAD

**Issue by leonardder** [nvaccess/nvda#7178]
### Steps to reproduce:
1. Go to the Windows 10 settings app
2. Type this in the python console:
    > nav.appModule.isBadUIAWindow=lambda hwnd: True
3. Focus the categories list
4. Try to use next/previous object navigation commands

### Expected behavior:
You should be able to navigate the list using object nav

### Actual behavior:
NVDA reports no next/previous element. However, if you do nav.parent.children in the python console, you'll realise that there are multiple list items

### Remarks
I do realise the str sounds a bit strange, why should one ever want to forcefully disable UIA altogether in the settings app? However, I've seen another app used in a corporate environment (IAccessible only), in which I saw very similar behaviour.
### System configuration:
NVDA version:
next-14047,eae024d0

NVDA Installed

Windows version:
Windows 10 Version 1703 (OS Build 15063.296)

### Other questions:

Does the issue still occur after restarting your PC?
Yes

Have you tried any other versions of NVDA?
No

CC @MichaelDCurran @josephsl

**Comment by jcsteh** [comment]
This is probably due to accNavigate NAVDIR_NEXT/PREVIOUS not being
implemented due to Microsoft's absurd statement that these are deprecated.
There's no fast, reliable way to find the next/previous child in this case.
I guess you could look at the parent's children and compare the current
child with each child until you find it, then walk next or previous, but
doing this for every request could be extremely problematic for
performance, especially in large containers. It's also not reliable because
comparing objects in IAccessible is not reliable. If you got the object
from the parent itself, you'd already have the child id, but that's fairly
rare except with object navigation. Still, I agree this brokenness is
happening more in the wild now. The question is how to fix it in a way that
isn't too hideous and doesn't break current functionality.

**Comment by leonardder** [comment]
> I guess you could look at the parent's children and compare the current
> child with each child until you find it, then walk next or previous, but
> doing this for every request could be extremely problematic for
> performance, especially in large containers. It's also not reliable because
> comparing objects in IAccessible is not reliable. If you got the object
> from the parent itself, you'd already have the child id, but that's fairly
> rare except with object navigation. Still, I agree this brokenness is
> happening more in the wild now. The question is how to fix it in a way that
> isn't too hideous and doesn't break current functionality.

I agree that the whole comparing and walking through all the children of a parent can be problematic. Here is another idea.

1. Enhance the indexINParent property on the IAccessible object to allow it returning something like self.IAccessibleChildID or self.parent.children.index(self). If that fails, return None
2. Add an additional indexInParent parameter to the IAccessible object constructor, which, if provided, overrides the indexINParent property. We can provide the indexInParent parameter in object construction for _get_firstChild, _get_lastChild, getChild and _get_children
3. If accNavigate returns nothing, try to use self.parent.getChild(self.indexInParent-|+1)

**Comment by jcsteh** [comment]
> I agree that the whole comparing and walking through all the children of a parent can be problematic. Here is another idea.
> 1. Enhance the indexINParent property on the IAccessible object to allow it returning something like self.IAccessibleChildID or self.parent.children.index(self).

`self.parent.children.index` is exactly walking the children and comparing. First, `self.parent.children` could be really expensive, especially in a large container. Second, `.index` will compare each item with `self`, and for IAccessible, comparison is unreliable/unperformant as I described earlier. This is the really problematic bit. The rest of what you suggest makes sense and is what I would do.

**Comment by leonardder** [comment]
>  self.parent.children.index  is exactly walking the children and comparing. First,  self.parent.children  could be really expensive, especially in a large container. Second,  .index  will compare each item with  self , and for IAccessible, comparison is unreliable/unperformant as I described earlier. This is the really problematic bit.

I initially proposed this point as a fallback meganism for when the indexInParent info wouldn't be available from other sources, such as the IAccessibleChildID or from the object's constructor as described in point 2.

I'll implement something based on point 2 and 3 and see how it works in the wild.

**Comment by jcsteh** [comment]
Note that IAccessibleChildID will always be 0 except for simple children
(elements that don't have their own IAccessible object). You may also want
to look at event_childID, which will at least allow this to work if the
object came from an event... in some cases. It won't work if the object
uses custom object ids or negative child ids to identify objects.

**Comment by leonardder** [comment]
Thanks for the clarification. I assume event_childID should be positive and that they are indexes starting at 1 if they are valid?

**Comment by jcsteh** [comment]
You should ignore values less than or equal to 0, yes. Unfortunately, there
are some implementations where a positive id is also used for a unique id
(rather than a child index) and these are going to break pretty horribly.
Hopefully, those implementations do implement accNavigate... but you never
know.

**Comment by leonardder** [comment]
I think we can avoid this horrible breakage by always checking whether the index is less than the child count

**Comment by jcsteh** [comment]
Oh, of course. Good thinking.

**Comment by leonardder** [comment]
Ok, i have something that is almost ready for a pr.

There is one issue though. The new meganism for next/previous is only used when indexInParent is known. However, this index is unknown when the navigator object is changed (e.g. by a focus change) or when you go to a parent element. So, the functionality does only work when you start from the top level window of the broken application and move downwards into that.

Any thoughts on how to solve this other than working with obj.parent.children.index and the broken IAccessible object comparison?

**Comment by jcsteh** [comment]
> There is one issue though. The new meganism for next/previous is only used when indexInParent is known. However, this index is unknown when the navigator object is changed (e.g. by a focus change) or when you go to a parent element.

That's exactly what I was pointing out [above](https://github.com/nvaccess/nvda/issues/7178#issuecomment-301624462):

> If you got the object from the parent itself, you'd already have the child id, but that's fairly rare except with object navigation.

> Any thoughts on how to solve this other than working with obj.parent.children.index and the broken IAccessible object comparison?

1. As noted [above](https://github.com/nvaccess/nvda/issues/7178#issuecomment-301987669), you can use event_childID in some cases if the object came from an event. Again, you should do the `0 < event_childID < childCount` test.
2. If `accNavigate` isn't implemented for next or previous, it's reasonable to assume it won't be implemented for firstChild or lastChild either. If the object was fetched from a parent and `accNavigate` failed to get the first or last child, the parent should probably be cached on the child. That way, you won't lose indexInParent. You'll probably have to tweak the IAccessible NVDAObject constructor to allow for this.
3. Otherwise, the only way is to fall back to the children fetch, which is the one that really bothers me. IMO, we should consider disabling this fallback by default (but providing an easy way to enable it in subclasses), at least initially.

**Comment by leonardder** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [17 mei 2017 07:15 CEST](https://github.com/nvaccess/nvda/issues/7178#issuecomment-301987669 "2017-05-17T05:15:04Z - Replied by Github Reply Comments"):
* It won't work if the object
> uses custom object ids or negative child ids to identify objects.

Yuck, this is exactly what happens if you use IAccessible for UIA objects.

I've been fiddling with storing the index in parent as well as the cached parent on the object, but this breaks horribly sometimes, leading to one object popping up multiple times into the object tree. Thus, this should be done if we are entirely sure that accNavigate isn't implemented. I think we can at least be pretty sure in the following situations:

1. obj.firstChild.next is None while obj.IAccessibleObject.accChildCount>1
2. obj.lastChild.previous is None while obj.IAccessibleObject.accChildCount>1

**Comment by Adriani90** [comment]
cc: @josephsl

**Comment by leonardder** [comment]
For refference, @jcsteh just noted on Twitter that Zoom also suffers from this:

> ... it returns E_NOTIMPL for IAccessible::accNavigate with NAVDIR_NEXT/PREVIOUS...

**Comment by LeonarddeR** [comment]
This has been abandoned for ages and apart from my work for Babbage, I haven't seen this elsewhere.

**Comment by jcsteh** [comment]
FWIW, Zoom still seems to suffer from this. But also, the Zoom desktop app is all kinds of broken with regard to accessibility, so maybe that just needs to be tackled as a Zoom specific hack (among others).

**Comment by LeonarddeR** [comment]
Does Zoom only offer MSAA or is it IAccessible2 or the like?

**Comment by jcsteh** [comment]
Only MSAA as far as I know.
